### PR TITLE
Add the ability to create and send a voting data

### PR DESCRIPTION
**Pull request by MichaelKim20**
I added information to access Agora to BOAClient
I added the ability to create and send a voting transaction
I have already done PR #39  on the issue.
However, after thinking more about it, I have created a new PR.

Related to #17

### CODE REVIEW

File: tests/Hash.test.ts [hunk 1]
```diff
@@ -82,4 +82,36 @@ describe('Hash', () => {
...
+            ],
+            [
+                new boasdk.TxOutput("0", new boasdk.PublicKey())
+            ]
+        );
+
+        assert.strictEqual(boasdk.hashFull(freeze_tx).toString(),
+            "0x7c8f238f70318aae788487e57eb6f4793e3b0e48fbe017c1829ec269d07a76661" +
+            "5173e9a33739e916b1d40248a48add0296b2a8e51d0f8e1a87bf7acc7c4f136");
+    });
```
Comment (MichaelKim20) [inline tests/Hash.test.ts:116]: I was able to achieve the same result as Agora.

File: src/modules/data/Transaction.ts [hunk 2]
```diff
@@ -47,13 +50,19 @@ export class Transaction
      */
     public outputs: TxOutput[];
 
+    /**
+     * The data to store
+     */
+    public data: SmartBuffer;
```
Comment (MichaelKim20) [inline src/modules/data/Transaction.ts:56]: This should be implemented in Agora.

Comment (MichaelKim20) [reply src/modules/data/Transaction.ts:56]: Use SmartBuffer because Buffer cannot expand its size.

File: src/modules/data/Transaction.ts [hunk 3]
```diff
@@ -82,5 +95,104 @@ export class Transaction
             elem.computeHash(buffer);
         for (let elem of this.outputs)
             elem.computeHash(buffer);
+
+        this.data.readOffset = 0;
+        buffer.writeBuffer(this.data.readBuffer())
```
Comment (MichaelKim20) [inline src/modules/data/Transaction.ts:100]: `this.data` was also added when calculating the hash of transactions.

File: src/modules/data/Transaction.ts [hunk 4]
```diff
@@ -82,5 +95,104 @@ export class Transaction
...
+    public static create (inputs: Array<TxInput>, outputs: Array<TxOutput>, keys: Array<Seed>, data: Buffer): Transaction
+    {
+        if (inputs.length == 0)
+            throw new Error ("The number of inputs is 0.");
+
+        if (outputs.length == 0)
+            throw new Error ("The number of outputs is 0.");
+
+        if (inputs.length != keys.length)
+            throw new Error ("The number of inputs and keys are different.");
```
Comment (MichaelKim20) [inline src/modules/data/Transaction.ts:120]: Validate parameters.

File: src/modules/data/Transaction.ts [hunk 5]
```diff
@@ -82,5 +95,104 @@ export class Transaction
...
+
+        let key_pairs: Array<KeyPair> = [];
+        for (let elem of keys)
+            key_pairs.push(KeyPair.fromSeed(elem));
```
Comment (MichaelKim20) [inline src/modules/data/Transaction.ts:124]: Create KeyPairs with seed keys.

File: src/modules/data/Transaction.ts [hunk 6]
```diff
@@ -82,5 +95,104 @@ export class Transaction
...
+
+        let tx = new Transaction(TxType.Payment, inputs, outputs, data);
+        let tx_hash = hashFull(tx);
+        for (let idx = 0; idx < tx.inputs.length; idx++)
+            tx.inputs[idx].signature.data.set(key_pairs[idx].secret.sign(tx_hash.data).data)
```
Comment (MichaelKim20) [inline src/modules/data/Transaction.ts:129]: Create a transaction and sign with a secret key.

File: src/modules/data/Transaction.ts [hunk 7]
```diff
@@ -82,5 +95,104 @@ export class Transaction
...
+
+        return tx;
     }
+
+    /**
+     * Gets the JSON object
+     * @returns The JSON object
+     * ex) {"type": 0, "inputs": inputs, "outputs": outputs, "data": data}
+     */
+    public toObject (): ITransaction
```
Comment (MichaelKim20) [inline src/modules/data/Transaction.ts:139]: Creates JSON data for transactions to be sent to Agora.

File: tests/BOAClient.test.ts [hunk 8]
```diff
@@ -558,4 +558,136 @@ describe ('BOA Client', () =>
...
+        ];
+
+        let vote_tx = boasdk.Transaction.create(
+            inputs,
+            outputs,
+            keys,
+            Buffer.from("vote data")
+        );
+
+        let expected_object = {
```
Comment (MichaelKim20) [inline tests/BOAClient.test.ts:612]: JSON data to be sent to Agora (be expected value)

File: tests/BOAClient.test.ts [hunk 9]
```diff
@@ -558,4 +558,136 @@ describe ('BOA Client', () =>
...
+        let obj = vote_tx.toObject();
+        assert.deepStrictEqual(obj, expected_object);
+
+        // Verify the signature
+        let tx_hash = boasdk.hashFull(vote_tx);
+        for (let idx = 0; idx < vote_tx.inputs.length; idx++)
+        {
+            let key_pair = boasdk.KeyPair.fromSeed(keys[idx]);
+            assert.ok(key_pair.address.verify(vote_tx.inputs[idx].signature, tx_hash.data));
+        }
```
Comment (MichaelKim20) [inline tests/BOAClient.test.ts:649]: This tests the verification of transactions signed with secret keys with public key.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**